preshed/maps.pyx

```diff
@@ -218,48 +205,44 @@ cdef bint map_iter(const MapStruct* map_, int* i, key_t* key, void** value) nogi
 
 
 @cython.cdivision
-cdef inline Cell* _find_cell(Cell* cells, const key_t size, const key_t key) nogil:
+cdef inline Cell _find_cell(const vector[Cell]& cells, const key_t key) nogil:
     # Modulo for powers-of-two via bitwise &
-    cdef key_t i = (key & (size - 1))
+    cdef key_t i = (key & (cells.size() - 1))
     while cells[i].key != EMPTY_KEY and cells[i].key != key:
-        i = (i + 1) & (size - 1)
-    return &cells[i]
+        i = (i + 1) & (cells.size() - 1)
+    return cells[i]
 
 
 @cython.cdivision
-cdef inline Cell* _find_cell_for_insertion(Cell* cells, const key_t size, const key_t key) nogil:
+cdef inline vector[Cell].iterator _find_cell_for_insertion(vector[Cell]* cells, const key_t key) nogil:
     """Find the correct cell to insert a value, which could be a previously
     deleted cell. If we cross a deleted cell and the key is in the table, we
     mark the later cell as deleted, and return the earlier one."""
-    cdef Cell* deleted = NULL
+    cdef vector[Cell].iterator deleted = cells.end()
     # Modulo for powers-of-two via bitwise &
-    cdef key_t i = (key & (size - 1))
-    while cells[i].key != EMPTY_KEY and cells[i].key != key:
-        if cells[i].key == DELETED_KEY:
-            deleted = &cells[i]
-        i = (i + 1) & (size - 1)
-    if deleted is not NULL:
-        if cells[i].key == key:
+    cdef key_t i = (key & (cells.size() - 1))
+    while deref(cells)[i].key != EMPTY_KEY and deref(cells)[i].key != key:
+        if deref(cells)[i].key == DELETED_KEY:
+            deleted = cells.begin() + i
+        i = (i + 1) & (cells.size() - 1)
+    if deleted != cells.end():
+        if deref(deleted).key == key:
```

Comment:
We could avoid `deref`ing by assigning to a local reference. But we have to ensure it is a reference, otherwise the complete vector gets copied on every call.

Reply:
Sadly Cython does not support this:

`c++ references cannot be declared use a pointer instead`